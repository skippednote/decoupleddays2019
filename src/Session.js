import React from "react";
import { navigate, Link } from "@reach/router";
import { UseSession } from "./useSessions";
import { Picture } from "./Picture";

function Session(props) {
  const session = new UseSession(props.id, ["body"]).getSessions();
  const isLoggedIn = window.localStorage.getItem("token") || false;

  const deleteSession = async () => {
    const token = window.localStorage.getItem("token");
    await fetch(`http://testing.dd:8083/api/node/session/${props.id}`, {
      method: "DELETE",
      headers: {
        Authorization: token
      }
    });

    navigate("/");
  };

  if (session) {
    return (
      <div>
        <h1>{session.title}</h1>
        <p>{session.field_room}</p>
        <p>{session.field_date && session.field_date}</p>
        <p>{session.field_track && session.field_track.name}</p>
        <p
          dangerouslySetInnerHTML={{
            __html: session.body && session.body.value
          }}
        />
        <ul>
          {session.field_speaker &&
            session.field_speaker.map(speaker => (
              <li key={session.id + speaker.field_name}>
                <p>{speaker.field_name}</p>
                {Picture(speaker)}
              </li>
            ))}
        </ul>
        {Boolean(isLoggedIn) ? (
          <>
            <button onClick={deleteSession}>Delete this session</button>
            <Link to={`/sessions/${session.id}/edit`}>Edit session</Link>
          </>
        ) : (
          <Link to="/login">Login</Link>
        )}
      </div>
    );
  }

  return <div>Loading...</div>;
}

export { Session };
