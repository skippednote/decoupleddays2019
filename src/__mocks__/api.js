import {
  sessions,
  categories,
  users,
  createSessionSuccess,
  deleteSessionSuccess
} from "../utils/fakes";

export const API = {
  getSessions: jest.fn(url => {
    if (url) {
      return sessions[0];
    } else {
      return sessions;
    }
  }),
  createSession: jest.fn(() => {
    return createSessionSuccess;
  }),
  updateSession: jest.fn(() => {
    return createSessionSuccess;
  }),
  deleteSession: jest.fn(() => {
    return deleteSessionSuccess;
  }),
  getCategories: jest.fn(() => {
    return categories;
  }),
  getUsers: jest.fn(() => {
    return users;
  })
};
