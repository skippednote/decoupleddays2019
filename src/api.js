import { useState, useEffect } from "react";
import axios from "axios";
import normalize from "json-api-normalize";
import Axios from "axios";
import * as Sentry from "@sentry/browser";

class APIFactory {
  constructor() {
    this.url = "";
    this.fields = null;
    this.defaultFields = [
      "id",
      "title",
      "created",
      "field_room",
      "field_date",
      "field_length",
      "field_speakers",
      "field_speakers.id",
      "field_speakers.user_picture.uri",
      "field_speakers.field_full_name",
      "field_speakers.field_company",
      "field_category.id",
      "field_category.name"
    ];
    this.baseURL = `http://sessions.lndo.site/api`;
  }

  getSessions(url = "", fields = []) {
    const [session, setSession] = useState(null);
    useEffect(() => {
      const cancelToken = axios.CancelToken;
      const source = cancelToken.source();
      try {
        const getS = async () => {
          const { data: response } = await axios.get(
            this._generateURL("node/session/" + url, true),
            {
              cancelToken: source.token
            }
          );
          let f = [...this.defaultFields];
          if (fields) {
            f = [...f, ...fields];
          }
          const normalizedData = normalize(response).get(f);
          setSession(normalizedData);
        };
        getS();
      } catch (error) {
        setSession("error");
        Sentry.withScope(scope => {
          scope.setExtras(error.message);
          Sentry.captureException(error);
        });
      }

      return () => source.cancel();
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    return session;
  }

  getCategories() {
    const [categories, setCategories] = useState(null);

    useEffect(() => {
      const cancelToken = axios.CancelToken;
      const source = cancelToken.source();
      try {
        const get = async () => {
          const { data } = await axios.get(
            this._generateURL("taxonomy_term/category")
          );

          setCategories(
            data.data.map(category => ({
              value: category.id,
              label: category.attributes.name
            }))
          );
        };

        get();
      } catch (error) {
        setCategories("error");
        Sentry.withScope(scope => {
          scope.setExtras(error.message);
          Sentry.captureException(error);
        });
      }

      return () => source.cancel();
    }, []);

    return categories;
  }

  getUsers() {
    const [users, setUsers] = useState(null);
    useEffect(() => {
      const cancelToken = axios.CancelToken;
      const source = cancelToken.source();
      try {
        const get = async () => {
          const { data } = await axios.get(this._generateURL("user/user"));
          setUsers(
            data.data.map(u => ({
              value: u.id,
              label: u.attributes.field_full_name || "Anonymous"
            }))
          );
        };

        get();
      } catch (error) {
        setUsers("error");
        Sentry.withScope(scope => {
          scope.setExtras(error.message);
          Sentry.captureException(error);
        });
      }

      return () => source.cancel();
    }, []);

    return users;
  }

  async createSession(values) {
    try {
      const data = this._generateSubmitData(values);
      const { data: session } = await axios.post(
        this._generateURL("node/session"),
        data,
        {
          headers: {
            Accept: "application/vnd.api+json",
            "Content-Type": "application/vnd.api+json",
            Authorization: window.localStorage.getItem("token")
          }
        }
      );
      return { id: session.data.id, error: false };
    } catch (error) {
      Sentry.withScope(scope => {
        scope.setExtras(error.message);
        Sentry.captureException(error);
      });
      return { id: false, error: true };
    }
  }

  async updateSession(values, id) {
    try {
      const data = this._generateSubmitData(values, id);
      const { data: session } = await axios.patch(
        this._generateURL("node/session/" + id),
        data,
        {
          headers: {
            Accept: "application/vnd.api+json",
            "Content-Type": "application/vnd.api+json",
            Authorization: window.localStorage.getItem("token")
          }
        }
      );
      return { id: session.data.id, error: false };
    } catch (error) {
      Sentry.withScope(scope => {
        scope.setExtras(error.message);
        Sentry.captureException(error);
      });
      return { id: false, error: true };
    }
  }

  async deleteSession(id) {
    try {
      const token = window.localStorage.getItem("token");
      await Axios.delete(this._generateURL(`node/session/${id}`), {
        headers: {
          Authorization: token
        }
      });

      return { error: false, success: true };
    } catch (error) {
      Sentry.withScope(scope => {
        scope.setExtras(error.message);
        Sentry.captureException(error);
      });
      return { error: true, success: false };
    }
  }

  _generateURL(subpath = "", include = false) {
    let queryParams = "";
    if (include) {
      queryParams = `?include=field_speakers,field_speakers.user_picture,field_category&sort=-created`;
    }
    return `${this.baseURL}/${subpath}${queryParams}`;
  }

  _generateSubmitData(values, id) {
    const data = {
      data: {
        type: "node--session",
        attributes: {
          title: values.title,
          body: {
            value: values.body,
            format: "full_html"
          },
          field_length: Number(values.slength),
          field_room: values.room,
          field_date: `${values.date}T${values.time}+00:00`
        },
        relationships: {
          field_category: {
            data: {
              type: "taxonomy_term--category",
              id: values.category
            }
          },
          field_speakers: {
            data:
              values.speakers &&
              values.speakers.map(speaker => ({
                type: "user--user",
                id: speaker.value
              }))
          }
        }
      }
    };

    if (id) {
      data.data.id = id;
      return data;
    }

    return data;
  }
}
const API = new APIFactory();

export { API };
