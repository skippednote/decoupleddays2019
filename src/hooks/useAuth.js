import { useState, useContext, useEffect, createContext } from "react";

export const AuthContext = createContext({});

function useAuth() {
  const [token, setToken] = useState(window.localStorage.getItem("token"));

  useEffect(() => {
    if (token) {
      window.localStorage.setItem("token", token);
    } else {
      window.localStorage.removeItem("token");
    }
  }, [token]);

  return { token, setToken };
}

export const useAuthContext = () => useContext(AuthContext);

export { useAuth };
