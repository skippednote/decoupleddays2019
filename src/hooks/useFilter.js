import { useState, useEffect } from "react";
import { getDate } from "../utils/getDate";

export function useFilter(sessions) {
  const [filtered, setFiltered] = useState(sessions);
  const [filter, setFilter] = useState({
    category: "All",
    date: "All",
    input: ""
  });

  useEffect(() => {
    const filteredSession = sessions.filter(session => {
      const { date: sessionDate } = getDate(session.field_date);
      return (
        (filter.category === "All" ||
          session.field_category.name === filter.category) &&
        (filter.date === "All" || sessionDate === filter.date) &&
        (filter.input === "" || session.title.indexOf(filter.input) > -1)
      );
    });

    setFiltered(filteredSession);
  }, [sessions, filter]);

  return { filter, setFilter, filtered };
}
