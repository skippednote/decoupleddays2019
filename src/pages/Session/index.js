import React, { useState, lazy, Suspense } from "react";
import { API } from "../../api";
const SessionItem = lazy(() => import("../../components/SessionItem"));

function Session({
  history,
  match: {
    params: { id }
  }
}) {
  const [status, setStatus] = useState({ id: null, error: null });
  const session = API.getSessions(id, ["body"]);

  const deleteHandler = async () => {
    const { error, success } = await API.deleteSession(session.id);

    setStatus({ error, success });

    if (success) {
      history.push("/");
    }
  };

  if (session) {
    return (
      <Suspense
        fallback={
          <div data-testid="loading-component">Loading component...</div>
        }
      >
        <SessionItem
          session={session}
          status={status}
          deleteHandler={deleteHandler}
        />
      </Suspense>
    );
  }

  return <div data-testid="loading-data">Loading data...</div>;
}

export default Session;
