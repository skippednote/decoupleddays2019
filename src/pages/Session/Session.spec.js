import React from "react";
import { render, wait, fireEvent } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Session from ".";
import { AuthContext } from "../../hooks/useAuth";
import { API as MockAPI } from "../../api";

const historyMock = { push: jest.fn() };
const matchMock = { params: { id: 3000 } };
jest.mock("../../api.js");

afterEach(() => {
  MockAPI.getSessions.mockClear();
  MockAPI.deleteSession.mockClear();
  historyMock.push.mockClear();
});

it("renders the component", async () => {
  const { container, getByTestId } = render(
    <MemoryRouter>
      <Session history={historyMock} match={matchMock} />
    </MemoryRouter>
  );

  expect(container).toMatchSnapshot();

  await wait(() => expect(getByTestId("session-item")).toBeInTheDocument());
  expect(container).toMatchSnapshot();
});

it("renders edit and delete button when logged in", async () => {
  const { getByText } = render(
    <MemoryRouter>
      <AuthContext.Provider value={{ token: "yo" }}>
        <Session history={historyMock} match={matchMock} />
      </AuthContext.Provider>
    </MemoryRouter>
  );

  await wait(() => {
    expect(getByText("Delete this session")).toBeInTheDocument();
    expect(getByText("Edit session")).toBeInTheDocument();
  });
});

it("calls the api", async () => {
  const { getByTestId } = render(
    <MemoryRouter>
      <Session history={historyMock} match={matchMock} />
    </MemoryRouter>
  );
  await wait(() => expect(getByTestId("session-item")).toBeInTheDocument());
  expect(MockAPI.getSessions).toHaveBeenCalledTimes(1);
});

it("redirects to homepage when session is deleted", async () => {
  const { getByText } = render(
    <MemoryRouter>
      <AuthContext.Provider value={{ token: "hello" }}>
        <Session history={historyMock} match={matchMock} />
      </AuthContext.Provider>
    </MemoryRouter>
  );

  await wait(() => getByText("Delete this session"));
  fireEvent.click(getByText("Delete this session"));
  expect(MockAPI.deleteSession).toHaveBeenCalledTimes(1);
  expect(MockAPI.deleteSession).toHaveBeenCalledWith(
    "017a6b73-fdcb-473f-aa49-8ae0073ab076"
  );
  await wait(() => {
    expect(historyMock.push).toHaveBeenCalled();
    expect(historyMock.push).toHaveBeenCalledWith("/");
  });
});

it("stays on the same page when deleting session fails", async () => {
  MockAPI.deleteSession.mockReset();
  MockAPI.deleteSession.mockImplementationOnce(() => ({
    error: true,
    status: false
  }));

  const { getByText } = render(
    <MemoryRouter>
      <AuthContext.Provider value={{ token: "hello" }}>
        <Session history={historyMock} match={matchMock} />
      </AuthContext.Provider>
    </MemoryRouter>
  );

  await wait(() => getByText("Delete this session"));
  fireEvent.click(getByText("Delete this session"));
  expect(MockAPI.deleteSession).toHaveBeenCalledTimes(1);
  expect(MockAPI.deleteSession).toHaveBeenCalledWith(
    "017a6b73-fdcb-473f-aa49-8ae0073ab076"
  );
  await wait(() => {
    expect(historyMock.push).not.toHaveBeenCalled();
  });
});

it("renders the loading message while waiting for API response", async () => {
  MockAPI.getSessions.mockReset();
  MockAPI.getSessions.mockImplementationOnce(() => {
    setTimeout(() => {}, 100);
  });
  const { getByTestId } = render(
    <MemoryRouter>
      <AuthContext.Provider value={{ token: "hello" }}>
        <Session history={historyMock} match={matchMock} />
      </AuthContext.Provider>
    </MemoryRouter>
  );
  expect(getByTestId("loading-data")).toBeInTheDocument();
});
