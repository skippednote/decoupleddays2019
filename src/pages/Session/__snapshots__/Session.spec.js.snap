// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<div>
  <div
    data-testid="loading-component"
  >
    Loading component...
  </div>
</div>
`;

exports[`renders the component 2`] = `
<div>
  <div
    data-testid="session-item"
  >
    <h4
      class="session-title"
      data-testid="session-title"
    >
      <a
        href="/017a6b73-fdcb-473f-aa49-8ae0073ab076"
      >
        a
      </a>
      <span
        class="session-length"
      >
        25
         minutes
      </span>
    </h4>
    <div
      class="session-meta"
    >
      <p
        class="session-room"
        data-testid="session-room"
      >
        room1
      </p>
      <span
        class="spacer"
      >
        •
      </span>
      <p
        class="session-date"
        data-testid="session-date"
      >
        2019-07-09T00:01:00+00:00
      </p>
      <span
        class="spacer"
      >
        •
      </span>
      <p
        class="session-category"
        data-testid="session-category"
      >
        Decoupled Drupal
      </p>
    </div>
    <p
      class="session-body"
    >
      <p>
        Hello body
      </p>
      

    </p>
    <hr />
    <ul
      class="session-speakers"
      data-testid="session-speakers"
    >
      <li>
        <p>
          Bassam Ismail
           
          (Axelerant)
        </p>
      </li>
    </ul>
  </div>
</div>
`;
